### PR TITLE
add balanced_image_dataset generator for triplets generation

### PR DESCRIPTION
**Pull request by ma7555**
This PR introduces a new image dataset generator that produces a specific number of classes and specific number of images per class in every batch.

This is needed for calculating some of the newer loss functions like:

* [TripletHardLoss](https://www.tensorflow.org/addons/api_docs/python/tfa/losses/TripletHardLoss)
* [TripletSemiHardLoss](https://www.tensorflow.org/addons/api_docs/python/tfa/losses/TripletSemiHardLoss)

Reason for this is that by randomly sampling from a dataset with relatively large number of classes (specially if balanced) and a small batch_size, it is highly unlikely to produce any positive pairs in most batches.

And since the loss is calculated by: [L191](https://github.com/tensorflow/addons/blob/8cec33fcaaf1cf90aec7bdd55a0fcdbb251ce5c2/tensorflow_addons/losses/triplet.py#L191)

```
    triplet_loss = tf.math.truediv(
        tf.math.reduce_sum(
            tf.math.maximum(tf.math.multiply(loss_mat, mask_positives), 0.0)
        ),
        num_positives,
    )
```
where `num_positives` is the number of positive pairs in the batch, this leads to a division by zero.

This dataset generator introduces `safe_triplet` argument, which when set to True ensures that all batches including later ones include the same number of classes per batch and images per class.

Weighted sampling is used in order to ensure the higest dataset utlization when safe_triplet is True.

This PR should solve the below issues:
* https://github.com/tensorflow/tensorflow/issues/30199
* https://github.com/tensorflow/tensorflow/issues/46123
* https://github.com/tensorflow/tensorflow/issues/24833
* https://github.com/tensorflow/addons/issues/1676
* https://github.com/tensorflow/addons/issues/2447
* https://github.com/tensorflow/addons/issues/1938

Questions on stackoverflow:
* [nan-loss-in-keras-with-triplet-loss](https://stackoverflow.com/questions/59857152/nan-loss-in-keras-with-triplet-loss)
* [loss-nan-for-triplet-loss-tensorflow](https://stackoverflow.com/questions/67240684/loss-nan-for-triplet-loss-tensorflow)
* [training-using-triplet-loss-nan-in-summary-histogram-tensorflow](https://stackoverflow.com/questions/59987004/training-using-triplet-loss-nan-in-summary-histogram-tensorflow)

For more info about the Triplet Loss:
* [TripletLoss Tutorial](https://www.tensorflow.org/addons/tutorials/losses_triplet)

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR!

1. The PR contains a lot of code that is redundant with other utilities. We should never be copy/pasting existing logic to create new components.
2. The API would need to be approved: `balanced_image_dataset` doesn't tell me what it's for (it's for triplet generation), etc.
3. This is not a good fit for core Keras but it may be a good fit for KerasCV, please create an issue on the repo and describe the API you would want to use. We'll discuss from there. https://github.com/keras-team/keras-cv

**Comment by ma7555** [comment]
Hello @fchollet,
Thanks for taking time to review the PR. 

1- Most used used functions are changed from existing logic, either inside functions or expected arguments. only `load_image` is exactly the same but requires both image_path and label as input. I agree with you that maybe this can be reused more efficiently.
2- balanced_image_dataset API produces balanced batches from images in a directory with equal number of samples from a specific number of classes in every batch.
3- I will create an issue there.

Thanks
